aws/rust-runtime/aws-config/src/lib.rs

```diff
@@ -950,6 +954,81 @@ mod loader {
         }
 
+        #[tokio::test]
+        async fn test_origin_programmatic() {
+            let _ = tracing_subscriber::fmt::try_init();
+            let loader = defaults(BehaviorVersion::latest())
+                .test_credentials()
+                .http_client(NeverClient::new())
+                .profile_name("custom")
+                .profile_files(
+                    #[allow(deprecated)]
+                    ProfileFiles::builder()
+                        .with_contents(
+                            #[allow(deprecated)]
+                            ProfileFileKind::Config,
+                            "[profile custom]\nendpoint_url = http://localhost:8989",
+                        )
+                        .build(),
+                )
+                .endpoint_url("http://localhost:1111")
+                .load()
+                .await;
+            assert_eq!(Origin::shared_config(), loader.get_origin("endpoint_url"));
+        }
+
+        #[tokio::test]
+        async fn test_origin_env() {
+            let _ = tracing_subscriber::fmt::try_init();
+            let env = Env::from_slice(&[("AWS_ENDPOINT_URL", "http://localhost:7878")]);
+            let loader = defaults(BehaviorVersion::latest())
+                .test_credentials()
+                .http_client(NeverClient::new())
+                .env(env)
+                .profile_name("custom")
+                .profile_files(
+                    #[allow(deprecated)]
+                    ProfileFiles::builder()
+                        .with_contents(
+                            #[allow(deprecated)]
+                            ProfileFileKind::Config,
+                            "[profile custom]\nendpoint_url = http://localhost:8989",
+                        )
+                        .build(),
+                )
+                .load()
+                .await;
+            assert_eq!(
+                Origin::shared_environment_variable(),
+                loader.get_origin("endpoint_url")
+            );
+        }
+
+        #[tokio::test]
+        async fn test_origin_fs() {
+            let _ = tracing_subscriber::fmt::try_init();
+            let loader = defaults(BehaviorVersion::latest())
+                .test_credentials()
+                .http_client(NeverClient::new())
+                .profile_name("custom")
+                .profile_files(
+                    #[allow(deprecated)]
+                    ProfileFiles::builder()
+                        .with_contents(
+                            #[allow(deprecated)]
+                            ProfileFileKind::Config,
+                            "[profile custom]\nendpoint_url = http://localhost:8989",
+                        )
+                        .build(),
+                )
+                .load()
+                .await;
+            assert_eq!(
+                Origin::shared_profile_file(),
+                loader.get_origin("endpoint_url")
+            );
+        }
+
```

Comment:
The above tests are mutually exclusive in terms of where a config value comes from, which is great to have. Should we also add tests verifying the expected origin gets returned when config values are specified in multiple sources?

Reply:
I did this for 2/3 tests [^1]. As for the profile file test, I'm not sure how I would except by stubbing out some IMDS thing since that's the only thing with a lower precedence.

[^1]: The `test_origin_env` and `test_origin_programmatic` tests also set profile config, which is then overridden.